pkg/auth/handler/sso/auth_handler.go

```diff
@@ -139,73 +139,72 @@ func (authHandlerError) Error() string {
 
 func (h AuthHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	_, err := handler.Transactional(h.TxContext, func() (interface{}, error) {
-		success := h.Handle(w, r)
-		if success {
+		err := h.Handle(w, r)
+		if err == nil {
 			return nil, h.HookProvider.WillCommitTx()
 		}
-		return nil, authHandlerError{}
+		return nil, err
 	})
 	if err == nil {
 		h.HookProvider.DidCommitTx()
 	}
 }
 
-func (h AuthHandler) Handle(w http.ResponseWriter, r *http.Request) (success bool) {
+func (h AuthHandler) Handle(w http.ResponseWriter, r *http.Request) error {
 	var ok interface{}
 	var oauthAuthInfo sso.AuthInfo
-	success = false
 
 	// We have to return error by directly writing to response at this stage
 	// because we do not have valid state.
 	if h.Provider == nil {
 		http.Error(w, "Provider is not supported", http.StatusBadRequest)
-		return
+		return authHandlerError{}
 	}
 
 	payload, err := h.DecodeRequest(r)
 	if err != nil {
 		http.Error(w, "Failed to decode request", http.StatusBadRequest)
-		return
+		return authHandlerError{}
 	}
 
 	if err = payload.Validate(); err != nil {
 		http.Error(w, "Failed to validate request", http.StatusBadRequest)
-		return
+		return authHandlerError{}
 	}
 
 	reqPayload := payload.(AuthRequestPayload)
 
 	state, err := h.Provider.DecodeState(reqPayload.State)
 	if err != nil {
 		http.Error(w, "Failed to decode state", http.StatusBadRequest)
-		return
+		return authHandlerError{}
 	}
 
 	if err = h.validateCallbackURL(h.OAuthConfiguration.AllowedCallbackURLs, state.CallbackURL); err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
-		return
+		return authHandlerError{}
 	}
 
 	// From now on, we must return response by respecting CallbackURL and UXMode.
-	defer func() {
-		success = err == nil
-		switch state.UXMode {
-		case sso.UXModeWebRedirect, sso.UXModeWebPopup:
-			err = h.handleSessionResp(w, r, state.UXMode, state.CallbackURL, ok, err)
-		case sso.UXModeMobileApp:
-			err = h.handleRedirectResp(w, r, state.CallbackURL, ok, err)
-		default:
-			success = false
-			http.Error(w, "Invalid UXMode", http.StatusBadRequest)
-		}
-	}()
-
 	oauthAuthInfo, err = h.getAuthInfo(reqPayload)
-	if err != nil {
-		return
+	if err == nil {
+		ok, err = h.handle(oauthAuthInfo, *state)
 	}
-	ok, err = h.handle(oauthAuthInfo, *state)
-	return
+
+	var handlerError error
+	if err == nil {
+		handlerError = authHandlerError{}
+	}
```

Comment:
The condition seems wrong.

Reply:
I'll drop this change if original code is ok.